File: src/cell.py
class Cell:
    def __init__(self, x, y, column, row, section):
        self.x = x
        self.y = y
        self.value = 0

        self.column = column
        self.row = row
        self.section = section

    def available_values(self):
        return self.column.available_values\
                .intersection(self.row.available_values)\
                .intersection(self.section.available_values)

    def set_value(self, value):
        self.value = value
        self.column.pop_value(value)
        self.row.pop_value(value)
        self.section.pop_value(value)

    def unset_value(self):
        self.column.push_value(self.value)
        self.row.push_value(self.value)
        self.section.push_value(self.value)
        self.value = 0

File: web/bin/cell.js
export class Cell {
    constructor(x, y, column, row, section) {
        this.x = x;
        this.y = y;
        this.column = column;
        this.row = row;
        this.section = section;
        this.value = 0;
    }
    availableValues() {
        return this.section.intersectValues(this.column, this.row);
    }
    setValue(value) {
        this.value = value;
        this.column.popValue(value);
        this.row.popValue(value);
        this.section.popValue(value);
    }
    unsetValue() {
        const value = this.value;
        this.column.pushValue(value);
        this.row.pushValue(value);
        this.section.pushValue(value);
        this.value = 0;
        return value;
    }
}
